### DISCUSSION TITLE
Custom pipeline components with runtime parameters

### DISCUSSION THREAD

**Discussion by paulbriton** [explosion/spaCy#6943] [Help: Coding & Implementations] [answered]
I'm currently using spaCy NER to detect person entities on a document. I've created a custom component to set a custom attribute on entities which match a list of terms. The goal of this component is to exclude person entities which matches some terms:
```python
@Language.factory('excluded')
class ExcludedComponent:
    def __init__(self, nlp, name, terms: list):
        self.matcher = PhraseMatcher(nlp.vocab, attr='LOWER')
        self.matcher.add('TERMS', list(nlp.tokenizer.pipe(terms)))
        Span.set_extension('is_excluded', default=False)

    def __call__(self, doc):
        spans = []
        for _, start, end in self.matcher(doc):
            entity = Span(doc, start, end, 'PERSON')
            entity._.set('is_excluded', True)
            spans.append(entity)
        doc.ents = list(doc.ents) + spans
        return doc
```
Not sure if it's the best approach here to do what I want, but it's working quite well so far. The main issue I have with this implementation is that I'm using spaCy along with fastApi, and the list of terms I'm passing to this component depends on runtime parameters. What I would like to do is to load the model and initialize component during application startup, and then be able to pass the list of terms later on.
Is there another approach that I can use here to do what I wanna do? Is it possible to initialize a component with some configuration and update it later? Am I completely wrong to use a component for this task?

**Comment by ines** [comment, marked as answer]
One option could be to add a method to your component that adds the terms and that you can call later on after the `nlp` object has been created. Depending on how your application works, you might want to make that method re-initialize the `PhraseMatcher`, though – otherwise, you'll just end up adding to the same matcher over and over.

```python
excluded = nlp.get_pipe("excluded")
# This would call ExcludedComponent.add_terms
excluded.add_terms(["hello", "word", "..."])
```

Another option would be to not make this functionality a pipeline component – for example, you could do implement it as a [method extension](https://spacy.io/usage/processing-pipelines/#custom-components-attributes) on the `Doc` and then run:

```python
doc = nlp("hello world!")
doc._.exclude(["hello", "world", "..."])
```

(Btw, for the best performance, make sure you're only re-creating the `PhraseMatcher` if your terms change. Depending on your application and how the terms are generated, you could also cache the most frequent `Doc` objects created for your patterns, so you don't have to re-tokenize everything on every request.)

**Reply by paulbriton** [reply]
Thanks a lot for your answer and thanks also for this great lib you're building! Keep up the good work :)

**Reply by paulbriton** [reply]
Here is my implementation of the first option:
```python
@Language.factory('excluded')
class ExcludedComponent:
    def __init__(self, nlp, name):
        self.nlp = nlp
        self.matcher = PhraseMatcher(self.nlp.vocab, attr='LOWER')
        Span.set_extension('is_excluded', default=False)

    def add_terms(self, terms: list):
        self.matcher = PhraseMatcher(self.nlp.vocab, attr='LOWER')
        self.matcher.add('TERMS', list(self.nlp.tokenizer.pipe(terms)))

    def __call__(self, doc):
        spans = []
        for _, start, end in self.matcher(doc):
            entity = Span(doc, start, end, 'PERSON')
            entity._.set('is_excluded', True)
            spans.append(entity)
        doc.ents = list(doc.ents) + spans
        return doc
```

**Comment by graue70** [comment]
If the list of terms changes for every call of the pipeline, I think it makes sense to add them as a parameter to the `__call__` method like this:

```python
@Language.factory('excluded')
class ExcludedComponent:
    ...

    def __call__(self, doc, terms):
        ...
        return doc
```

This changes the calling code from
```python
excluded.add_terms(terms)
doc = nlp(input_text)
```
to
```python
doc = nlp(input_text, {"excluded": {"terms": terms}})
```
Note that the key of the outer dict must match the name of the component and the key of the inner dict must match the name of the method parameter.

This way you get an error if the terms aren't provided which prevents incorrect usage. On the other hand, you lose static type checking.
